vscode/src/chat/agentic/CodyToolProvider.ts

```diff
@@ -1,159 +1,201 @@
- * - Initializes and manages default Cody tools
- * - Manages OpenContext tools for external integrations
- * - Provides a unified interface to access all available tools
+ * Responsibilities:
+ * - Maintains a registry of tool configurations
+ * - Creates tool instances on demand
+ * - Handles both default tools (Search, File, CLI, Memory) and OpenCtx tools
+ * - Manages tool configuration and instantiation with proper context
  */
-export class CodyToolProvider {
-    private openCtxTools: CodyTool[] = []
-    private toolFactory = new ToolFactory()
-    private shellConfig: CodyShellConfig = {
-        user: false,
-        instance: false,
-        client: false,
-    }
+class ToolFactory {
+    private tools: Map<string, ToolConfiguration> = new Map()
 
-    private constructor(private contextRetriever: Pick<ContextRetriever, 'retrieveContext'>) {
-        this.initializeToolRegistry()
-        this.initializeOpenCtxTools()
+    constructor(private contextRetriever: Retriever) {
+        // Register default tools
+        for (const [name, { tool, useContextRetriever }] of Object.entries(TOOL_CONFIGS)) {
+            this.register({
+                name,
+                ...tool.prototype.config,
+                createInstance: useContextRetriever
+                    ? (_, contextRetriever) => {
+                          if (!contextRetriever) {
+                              throw new Error(`Context retriever required for ${name}`)
+                          }
+                          return new tool(contextRetriever)
+                      }
+                    : () => new tool(),
+            })
+        }
     }
 
-    public static instance(
-        contextRetriever: Pick<ContextRetriever, 'retrieveContext'>
-    ): CodyToolProvider {
-        return new CodyToolProvider(contextRetriever)
+    public register(toolConfig: ToolConfiguration): void {
+        this.tools.set(toolConfig.name, toolConfig)
     }
 
-    private initializeToolRegistry(): void {
-        registerDefaultTools(this.toolFactory.registry)
+    public createTool(name: string, retriever?: Retriever): CodyTool | undefined {
+        const config = this.tools.get(name)
+        if (!config) {
+            return undefined
+        }
+        const instance = config.createInstance(config, retriever)
+        return instance
     }
 
-    public setShellConfig(config: CodyShellConfig): void {
-        // merge the new config into the old config
-        const newConfig = { ...this.shellConfig, ...config }
-        this.shellConfig = newConfig
+    public getInstances(): CodyTool[] {
+        // Create fresh instances of all registered tools
+        return Array.from(this.tools.entries())
+            .filter(([name]) => name !== 'CliTool' || toolboxManager.getSettings()?.shell?.enabled)
+            .map(([_, config]) => config.createInstance(config, this.contextRetriever))
+            .filter(isDefined) as CodyTool[]
```

Comment:
With the type guard, we can remove type casting, making this code more type-safe

```suggestion
            .filter(isDefined)
```